### PR TITLE
[Serializer] Add delimiter auto detection for CSV decoding

### PR DESCRIPTION
**Pull request by welcoMattic**
| Q             | A
| ------------- | ---
| Branch?       | 6.2 
| Bug fix?      | no
| New feature?  | yes 
| Deprecations? | no 
| Tickets       | 
| License       | MIT
| Doc PR        | symfony/symfony-docs#... 

Recently I had an edge case about decoding CSV content for which I don't know the delimiter, it could be `,` or `;`.

I have implemented a simple guesser which loops over the headers row to guess the delimiter, based on a static list of popular CSV delimiters (`,`, `;`, `\t`, `|`).

I know that is a very particular case, but I guess if I needed it, may be someone will also do.

NOTE: it's not enable by default, only with `CsvEncoder::AUTO_DETECT_DELIMITER` set to `true`.

### CODE REVIEW

File: src/Symfony/Component/Serializer/Encoder/CsvEncoder.php [hunk 1]
```diff
@@ -292,4 +305,20 @@ private function extractHeaders(iterable $data): array
 
         return $headers;
     }
+
+    private function detectDelimiter(string $data): ?string
+    {
+        $headers = explode("\n", $data)[0];
```
Comment (dunglas) [inline src/Symfony/Component/Serializer/Encoder/CsvEncoder.php]:
```suggestion
        $headers = explode("\n", $data, 1)[0];
```

**Review by dunglas — COMMENTED** [review]
Nitpicking

**Comment by dunglas** [comment]
This method isn't very reliable right? It works only if there is a header line and if the headers don't contain too many special chars? I mean that with `f,,,,,oo;bar`, the helper will think that `,` is the delimiter while in reality it's `;`.
With no headers at all it's worse because data will likely be user-generated and this could lead to a security issue.

If we introduce this helper, I think that we must explicitly discourage to use it in the documentation, explain the possible reliability and security issues, and recommend to always use the `,` standard delimiter instead.

**Comment by fabpot** [comment]
> This method isn't very reliable right? It works only if there is a header line and if the headers don't contain too many special chars? I mean that with `f,,,,,oo;bar`, the helper will think that `,` is the delimiter while in reality it's `;`. With no headers at all it's worse because data will likely be user-generated and this could lead to a security issue.
> 
> If we introduce this helper, I think that we must explicitly discourage to use it in the documentation, explain the possible reliability and security issues, and recommend to always use the `,` standard delimiter instead.

This makes me think that it shouldn't be in core. :-1: for me.

**Comment by welcoMattic** [comment]
Thanks @dunglas and @fabpot for your comments, you are right, it's too edgy to be in the core. I close

**Pull request closed without merging**